### DISCUSSION TITLE
Testing form schemas (and possibly tables also)

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#6884] [Ideas]
A while back I did a bit of work on the testing for `Notifications`. One thing I liked about this was that in order to test that your notification was what was expected was that you could just do the following:
```php
...
->assertNotified(
    Notification::make('my_notification')
        ->title('My notification')
        ->success()
  )
```
This got me thinking that the same functionality would be great for forms (and possibly tables as well). When we need to check that a field is `required` for example, we currently have to do the following:
```php
livewire(CreateSomeResource::class)
    ->fillForm([
        'some_required_field' => ''
    ])
    ->assertFormHasErrors([
        'some_required_field' => 'required'
    ]);
```
This works and is fine but it feels like we're testing our own requirements as well as Filaments own implementation. It's also difficult to test that a field exists when it has no special requirements:
```
...
$form->schema([
    Toggle::make('some_boolean')
]);
```
With this in mind, the following seems to work for me and wanted to put it out there to see if a) there are any potential issues with this approach and b) whether this could this be wrapped up into an assertion (`assertFormSchema($schema)` ):

```php
use Filament\Resources\Form;
use App\Filament\Resources\MyResource;

$form = Form::make();
// the first item is a grid so we need the child comps of that
$schema = collect(MyResource::form($form)->getSchema())->first()->getChildComponents();

expect($schema)->toEqualCanonicalizing([
        TextInput::make('name')
            ->required(),
        TextInput::make('friendly_url')
            ->required()
            ->unique(ignoreRecord: true),
        RichEditor::make('full_description')
            ->disableToolbarButtons([
                'attachFiles',
                'codeBlock',
                'blockquote'
            ]),
        Toggle::make('is_featured'),
        Toggle::make('is_active')
            ->default(true)
]);
```
This is definately a sketch at the moment and know there needs to be more work to make this viable for both admin and forms as well different modes (editing and creating - a cursory issue is getting the resolved form from these resources as the methods are `protected`).

For me, this would drastically reduce the number of tests needed to ensure my forms are as they are expected and only test that my schema is correct. Also (but I need to check) this *should* improve test coverage as the static function `form` in the resource (and edit and create record classes) could now be covered.

**Comment by craigkuhns** [comment]
You mentioned that it's difficult to test that a field exists when doesn't have any requirements but in case you missed it there is an `assertFormFieldExists()` testing helper that can be used to ensure a field exists on a form and even allows accessing a given field in a closure to test it passes specific conditions. It's documented here https://filamentphp.com/docs/2.x/forms/testing#fields. That doesn't speak to your larger idea of simplifying form testing but it might make your life a bit easier for now.

**Comment by danharrin** [comment]
To be honest, I don't understand the value of tests that just duplicate the code that you're testing?

**Comment by titantwentyone** [comment]
@craigkuhns Thanks Craig. This one passed me by (or I forgot about it!)

@danharrin. The same could be said of the Notifications testing. As far as the value of it, it is my expectation that a form is set up in  certain way. If this changes, the test would fail - just like any other test. Particularly useful if working in a team where one developer feels something is redundant/irrelevant but a verifiable form schema says otherwise. It does replicate the code but it is way more succinct than testing every method attached to a field - especially when, in the example of `required()` it feels more like testing Filament rather than my own code. Should we test field methods that accept a closure or a custom field method in addition to this? Absolutely - this is no substitute for that.

For the following simple field:

```
TextInput::make('my_text')
    ->required()
    ->default('some text')
    ->minLength(2)
    ->maxLength(255);
```
To fully ensure this is as expected, we'd need, at minimum, four assertions. - which mostly test Filaments ability to fulfill its function. Asserting the form schema, just requires one.

**Reply by danharrin** [reply]
If you want to work on it, sure, but I still don't think I agree it actually has a purpose besides duplicating code. If you're worried about people removing code, thats what code review is for, right?
